issue_5/contracts/GreenVault.sol

```diff
@@ -1,93 +1,172 @@
+    @dev price feed
+    */
+    AggregatorV3Interface immutable priceFeed;
+
+    /**
+    @dev per-user data
+     */
+    struct UserData {
+        /**
+        @dev how much ETH was deposited
+         */
+        uint256 deposited;
+        /**
+        @dev how much ETH was borrowed against
+         */
+        uint256 borrowed;
+    }
 
-    It's represented as a wad (a 18-digit fixed point number)
+    /**
+    @dev Mapping (user address -> user data)
      */
-    uint256 public exchangeRateWad;
+    mapping(address => UserData) public users;
 
     /// @notice emitted on deposit
-    event Minted(address user, address token, uint256 amount);
-    /// @notice emitted on withdrawal
-    event Burned(address user, address token, uint256 amount);
+    event Deposited(uint256 amount);
 
-    event ExchangeRateSet(uint256 exchangeRateWad);
+    /// @notice emitted on withdrawal
+    event Borrowed(uint256 ethAmount, uint256 daiAmount);
+    /// @notice emitted when a debt is paid back
+    event PaidBack(uint256 ethAmount, uint256 daiAmount);
+    /// @notice emitted when the user withdraws fuunds
+    event Withdrawn(uint256 amount);
+    /// @notice emitted when the user is liquidated
+    event Liquidated(address user);
 
     /**
-    @param token address of the only token this vault will support
-    @param _exchangeRateWad exchange rate to start with
+    @param token address of DAI token
+    @param _priceFeed price feed address
      */
-    constructor(IERC20Metadata token, uint256 _exchangeRateWad) {
+    constructor(IERC20Metadata token, AggregatorV3Interface _priceFeed) {
         theOnlySupportedToken = token;
-        setExchangeRate(_exchangeRateWad);
+        priceFeed = _priceFeed;
     }
 
-    function setExchangeRate(uint256 _exchangeRateWad) public onlyOwner {
-        exchangeRateWad = _exchangeRateWad;
-        emit ExchangeRateSet(exchangeRateWad);
+    function getExchangeRate() internal view returns (uint256 rate, uint8 rateDecimals) {
+        (uint80 roundId, int256 answer, , , uint80 answeredInRound) = priceFeed
+            .latestRoundData();
+        require(answeredInRound == roundId, "Stale feed");
+        require(answer > 0, "negative exchange rate");
+        rate = (uint256)(answer);
+        rateDecimals = priceFeed.decimals();
     }
 
     /**
-    @notice Mint Vault token by depositing external token
-    @param amount amount of tokens to deposit
+    @notice Deposit ETH
      */
-    function deposit(uint256 amount) public {
+    function deposit() public payable {
+        users[msg.sender].deposited += msg.value;
+
+        emit Deposited(msg.value);
+    }
+
+    /**
+    @notice Borrow DAI against previously deposited ETH
+    @param collateral how much ETH to borrow against
+     */
+    function borrow(uint256 collateral) public {
         IERC20 token = theOnlySupportedToken;
+        UserData storage user = users[msg.sender];
```

Comment:
If you use `storage` the data storage becomes permanent, you should be using `memory` to work with temporary pointers to data. I don't think I've ever used `storage`.

Reply:
Today I learned! Disregard this comment until we get to gas optimization, then.